### PR TITLE
Update `tests/callbacks/*.py` to use `devices` instead of `gpus` or `ipus`

### PR DESCRIPTION
**Pull request by mathemusician**
## What does this PR do?

Update test to deprecate `num_processes`, `gpus`, `tpu_cores`, and `ipus` from the Trainer constructor

Part of [Pull 11040](https://github.com/PyTorchLightning/pytorch-lightning/pull/11040).

🌳 [Trello Board Link](https://trello.com/b/He49BfCc/pl-tests)

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/callbacks/test_xla_stats_monitor.py [hunk 1]
```diff
@@ -69,7 +77,7 @@ def test_xla_stats_monitor_no_tpu_warning(tmpdir):
     model = BoringModel()
     xla_stats = XLAStatsMonitor()
 
-    trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1, tpu_cores=None)
+    trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1, accelerator="tpu", devices=None)
```
Comment (kaushikb11) [inline tests/callbacks/test_xla_stats_monitor.py:80]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1, devices=None)
```

File: tests/callbacks/test_pruning.py [hunk 2]
```diff
@@ -185,7 +183,7 @@ def test_pruning_callback_ddp_spawn(tmpdir):
 
 @RunIf(skip_windows=True, skip_49370=True)
 def test_pruning_callback_ddp_cpu(tmpdir):
-    train_with_pruning_callback(tmpdir, parameters_to_prune=True, strategy="ddp_spawn", num_processes=2)
+    train_with_pruning_callback(tmpdir, parameters_to_prune=True, strategy="ddp_spawn", devices=2)
```
Comment (kaushikb11) [inline tests/callbacks/test_pruning.py:186]:
```suggestion
    train_with_pruning_callback(tmpdir, parameters_to_prune=True, strategy="ddp_spawn", accelerator="cpu", devices=2)
```

**Review by ananthsub — APPROVED** [review]

File: tests/callbacks/test_quantization.py [hunk 3]
```diff
@@ -43,7 +43,8 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     """Parity test for quant model."""
     seed_everything(42)
     dm = RegressDataModule()
-    trainer_args = dict(default_root_dir=tmpdir, max_epochs=7, gpus=int(torch.cuda.is_available()))
+    accelerator = "gpu" if torch.cuda.is_available() else "cpu"
+    trainer_args = dict(default_root_dir=tmpdir, max_epochs=7, accelerator=accelerator, devices=1)
```
Comment (carmocca) [inline tests/callbacks/test_quantization.py:47]:
```suggestion
    trainer_args = dict(default_root_dir=tmpdir, max_epochs=7, accelerator="auto", devices="auto")
```

File: tests/callbacks/test_xla_stats_monitor.py [hunk 4]
```diff
@@ -69,7 +77,7 @@ def test_xla_stats_monitor_no_tpu_warning(tmpdir):
     model = BoringModel()
     xla_stats = XLAStatsMonitor()
 
-    trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1, tpu_cores=None)
+    trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1, devices=None)
```
Comment (carmocca) [inline tests/callbacks/test_xla_stats_monitor.py:80]:
```suggestion
    trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1)
```

Isn't this the same?

**Review by carmocca — APPROVED** [review]

File: tests/callbacks/test_quantization.py (hunk 3 above)
Comment (ananthsub) [reply tests/callbacks/test_quantization.py:47]: devices=auto would do multi-gpu training, which is different from the current test behavior

File: tests/callbacks/test_xla_stats_monitor.py (hunk 4 above)
Comment (four4fish) [reply tests/callbacks/test_xla_stats_monitor.py:80]: I think this test doesn't make sense anymore, this scenario is not equally reproducible by `accelerator=` and `device=` 

This actually equal to 
`trainer = Trainer(default_root_dir=tmpdir, callbacks=[xla_stats], max_steps=1, device=None, accelerator="tpu")
But current accelerator_connector logic could handle it, if device =None, the device it will been set by accelerator_type(in `_set_devices_if_none`), and accelerator_type set by select_accelerator() which will check either accelerator flag or TPU availability. So the it won't be a Misconfig as the training will run on TPU.

File: tests/callbacks/test_quantization.py (hunk 3 above)
Comment (mathemusician) [reply tests/callbacks/test_quantization.py:47]: So don't change the code? Or do we want to use "auto"?

Comment (rohitgr7) [reply tests/callbacks/test_quantization.py:47]: I think just `accelerator='auto'` would be fine?

File: tests/callbacks/test_xla_stats_monitor.py (hunk 4 above)
Comment (mathemusician) [reply tests/callbacks/test_xla_stats_monitor.py:80]: Yeah, there seem to be some tests that won't `Misconfigure` when only using `devices` and `accelerator`. Should I write up a new test to get a Misconfiguration warning? Or should we delete these tests?

File: tests/callbacks/test_stochastic_weight_avg.py [hunk 5]
```diff
@@ -139,22 +145,22 @@ def train_with_swa(
...
 
 @RunIf(min_gpus=2)
 def test_swa_callback_ddp_spawn(tmpdir):
-    train_with_swa(tmpdir, strategy="ddp_spawn", gpus=2)
+    train_with_swa(tmpdir, strategy="ddp_spawn", accelerator="gpu", devices=2)
 
 
 @RunIf(skip_windows=True, skip_49370=True)
 def test_swa_callback_ddp_cpu(tmpdir):
-    train_with_swa(tmpdir, strategy="ddp_spawn", num_processes=2)
```
[the remark this answers is not in the document]
Comment (mathemusician) [reply tests/callbacks/test_stochastic_weight_avg.py:152]: Thanks for catching that! I'll use `grep` next time to make sure I catch them all. Using the regular "Find" doesn't do the trick.

File: tests/callbacks/test_quantization.py (hunk 3 above)
Comment (four4fish) [reply tests/callbacks/test_quantization.py:47]: `accelerator='auto'` is testing different things, current logic will test multi process CPU or single gpu, `accelerator='auto'`  will enable this test on multiple gpus . I think the Strategy will be different(not singleDevice anymore) which means the quant_score will be different as it need call collective to get all the tensors

The goal for this test is not to test the accelerator_connector flags, it's to test quantization, I think we should have flag reproduce test on cpu or single gpu

` ``
accelerator = "gpu" if torch.cuda.is_available() else "cpu"
 if accelerator == "gpu":
    trainer_args = dict(default_root_dir=tmpdir, max_epochs=7, accelerator=accelerator, devices=1)
 if accelerator == "cpu":
   trainer_args = dict(default_root_dir=tmpdir, max_epochs=7, accelerator=accelerator)
```
This should be the exactly reproduce.

Comment (four4fish) [reply tests/callbacks/test_quantization.py:47]: It's slightly different compared to before, but I don't think it impacts the accuracy of this tests

the reason we can not use 'auto': accelerator='auto' is testing different things, current logic will test multi process CPU or single gpu, accelerator='auto' will enable this test on multiple gpus . I think the Strategy will be different(not singleDevice anymore) which means the quant_score will be different as it need call collective to get all the tensors

Comment (mathemusician) [reply tests/callbacks/test_quantization.py:47]: Ah, so I shouldn't change the code because "auto" will do multi-gpu testing, but this test was made for one gpu. Right?

File: tests/callbacks/test_stochastic_weight_avg.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mathemusician) [reply tests/callbacks/test_stochastic_weight_avg.py:152]: Actually, the accelerator is already set to use "cpu". But I'll make that more explicit by adding `accelerator="cpu"`

[the remark this answers is not in the document]
Comment (four4fish) [reply tests/callbacks/test_stochastic_weight_avg.py:152]: It use cpu because the test machine (your laptop?) only have cpu, if this test run in ci and has GPUs or TPUs it will choose these over CPU. I think so, we need specific accelerator="cpu"

[the remark this answers is not in the document]
Comment (mathemusician) [reply tests/callbacks/test_stochastic_weight_avg.py:152]: The function is already set to use cpu, see [this line](https://github.com/mathemusician/pytorch-lightning/blob/944d6772f28bbc1c4ea8ce38431e95de04faafef/tests/callbacks/test_stochastic_weight_avg.py#L114)

File: tests/callbacks/test_xla_stats_monitor.py (hunk 4 above)
Comment (mathemusician) [reply tests/callbacks/test_xla_stats_monitor.py:80]: I'm leaning towards deleting the test entirely.

Comment (daniellepintz) [reply tests/callbacks/test_xla_stats_monitor.py:80]: I think we can delete it

Comment (awaelchli) [reply tests/callbacks/test_xla_stats_monitor.py:80]: nuke it. the name and description of the test is anyway lying.

**Review by awaelchli — APPROVED** [review]

**Comment by daniellepintz** [comment]
Can someone please rerun the `build-Docs` test in `ci-tests`?

**Review by akihironitta — APPROVED** [review]
